### PR TITLE
Eager print

### PR DESCRIPTION
**Pull request by nikste**
Uses collect in print function now.
No need for calling execute() after print().

### CODE REVIEW

**Comment by StephanEwen** [comment]
Okay, the issue with print in test cases seems to be fixed in the second commit, that looks good.

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 1]
```diff
@@ -1336,11 +1335,17 @@ public long count() throws Exception {
...
 	 * Writes a DataSet to the standard output stream (stdout).<br/>
 	 * For each element of the DataSet the result of {@link Object#toString()} is written.
-	 * 
-	 *  @return The DataSink that writes the DataSet.
+	 * This triggers execute() automatically.
 	 */
-	public DataSink<T> print() {
-		return output(new PrintingOutputFormat<T>(false));
+	public void print() throws Exception{
+		try {
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: How about removing this try/catch block?

We are getting a very deep nesting of exceptions already. Not sure this wrapping exception adds any information. Why not let the "ProgramInvokationFailed" exception from the execution propagate?

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 2]
```diff
@@ -1357,11 +1362,16 @@ public long count() throws Exception {
 	/**
 	 * Writes a DataSet to the standard error stream (stderr).<br/>
 	 * For each element of the DataSet the result of {@link Object#toString()} is written.
-	 * 
-	 * @return The DataSink that writes the DataSet.
 	 */
-	public DataSink<T> printToErr() {
-		return output(new PrintingOutputFormat<T>(true));
+	public void printToErr() throws Exception{
+		try {
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: See comment above

File: flink-optimizer/pom.xml [hunk 3]
```diff
@@ -58,6 +58,12 @@ under the License.
 			<artifactId>guava</artifactId>
 			<version>${guava.version}</version>
 		</dependency>
+		<dependency>
```
Comment (StephanEwen) [inline flink-optimizer/pom.xml]: This dependency should be removed, it should be unnecessary. If it is actually necessary, then there is an issue somewhere else ;-)

File: flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/CustomPartitioningGlobalOptimizationTest.java [hunk 4]
```diff
@@ -57,8 +58,8 @@ public void testJoinReduceCombination() {
 				.withPartitioner(partitioner);
 
 			joined.groupBy(1).withPartitioner(partitioner)
-				.reduceGroup(new IdentityGroupReducerCombinable<Tuple3<Long,Long,Long>>())
-				.print();
+				.reduceGroup(new IdentityGroupReducer<Tuple3<Long,Long,Long>>())
```
Comment (StephanEwen) [inline flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/CustomPartitioningGlobalOptimizationTest.java]: The semantics of the tests should not be changed. This needs to be changed back to the original `IdentityGroupReducerCombinable`

File: flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/GroupingKeySelectorTranslationTest.java [hunk 5]
```diff
@@ -84,8 +85,8 @@ public void testCustomPartitioningKeySelectorGroupReduce() {
 			
 			data.groupBy(new TestKeySelector<Tuple2<Integer,Integer>>())
 				.withPartitioner(new TestPartitionerInt())
-				.reduceGroup(new IdentityGroupReducerCombinable<Tuple2<Integer,Integer>>())
-				.print();
+				.reduceGroup(new IdentityGroupReducer<Tuple2<Integer,Integer>>())
```
Comment (StephanEwen) [inline flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/GroupingKeySelectorTranslationTest.java]: Also here, change back to the original function.

File: flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/GroupingKeySelectorTranslationTest.java [hunk 6]
```diff
@@ -115,8 +116,8 @@ public void testCustomPartitioningKeySelectorGroupReduceSorted() {
 			data.groupBy(new TestKeySelector<Tuple3<Integer,Integer,Integer>>())
 				.withPartitioner(new TestPartitionerInt())
 				.sortGroup(new TestKeySelector<Tuple3<Integer, Integer, Integer>>(), Order.ASCENDING)
-				.reduceGroup(new IdentityGroupReducerCombinable<Tuple3<Integer,Integer,Integer>>())
-				.print();
+				.reduceGroup(new IdentityGroupReducer<Tuple3<Integer,Integer,Integer>>())
```
Comment (StephanEwen) [inline flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/GroupingKeySelectorTranslationTest.java]: Also here, change back to the original function.

File: flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/GroupingPojoTranslationTest.java [hunk 7]
```diff
@@ -176,19 +177,19 @@ public void testCustomPartitioningTupleInvalidType() {
...
+
 	@Test
 	public void testCustomPartitioningTupleInvalidTypeSorted() {
 		try {
 			ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
-			
+
 			DataSet<Pojo3> data = env.fromElements(new Pojo3())
 					.rebalance().setParallelism(4);
-
```
Comment (StephanEwen) [inline flink-optimizer/src/test/java/org/apache/flink/optimizer/custompartition/GroupingPojoTranslationTest.java]: Many whitespace reformattings here. If you use IntelliJ, please set it to not auto-trim whitespaces. It blows up the diffs and distracts from actual changes.

File: flink-optimizer/src/test/java/org/apache/flink/optimizer/java/PartitionOperatorTest.java [hunk 8]
```diff
@@ -49,8 +50,8 @@ public void testPartitionOperatorPreservesFields() {
 					public int partition(Long key, int numPartitions) { return key.intValue(); }
 				}, 1)
 				.groupBy(1)
-				.reduceGroup(new IdentityGroupReducerCombinable<Tuple2<Long,Long>>())
-				.print();
+				.reduceGroup(new IdentityGroupReducer<Tuple2<Long,Long>>())
```
Comment (StephanEwen) [inline flink-optimizer/src/test/java/org/apache/flink/optimizer/java/PartitionOperatorTest.java]: Also here, change back to the original function.

File: flink-examples/flink-java-examples/src/main/java/org/apache/flink/examples/java/wordcount/WordCount.java [hunk 9]
```diff
@@ -73,13 +73,14 @@ public static void main(String[] args) throws Exception {
 
 		// emit result
 		if(fileOutput) {
-			counts.writeAsCsv(outputPath, "\n", " ");
+			counts.writeAsText(outputPath);
```
Comment (StephanEwen) [inline flink-examples/flink-java-examples/src/main/java/org/apache/flink/examples/java/wordcount/WordCount.java]: Can you roll this back to "writeAsCsv"?

File: flink-tests/src/test/java/org/apache/flink/test/exampleJavaPrograms/WordCountITCase.java [hunk 10]
```diff
@@ -36,7 +36,7 @@ protected void preSubmit() throws Exception {
 
 	@Override
 	protected void postSubmit() throws Exception {
-		compareResultsByLinesInMemory(WordCountData.COUNTS, resultPath);
+		compareResultsByLinesInMemory(WordCountData.COUNTS_AS_TUPLES, resultPath);
```
Comment (StephanEwen) [inline flink-tests/src/test/java/org/apache/flink/test/exampleJavaPrograms/WordCountITCase.java]: Followup of adjusted WordCount example

File: flink-staging/flink-tez/src/test/java/org/apache/flink/tez/test/WordCountITCase.java [hunk 11]
```diff
@@ -37,7 +37,7 @@ protected void preSubmit() throws Exception {
 
     @Override
     protected void postSubmit() throws Exception {
-        compareResultsByLinesInMemory(WordCountData.COUNTS, resultPath);
+        compareResultsByLinesInMemory(WordCountData.COUNTS_AS_TUPLES, resultPath);
```
Comment (StephanEwen) [inline flink-staging/flink-tez/src/test/java/org/apache/flink/tez/test/WordCountITCase.java]: Followup of adjusted WordCount example

File: flink-staging/flink-tachyon/src/test/java/org/apache/flink/tachyon/TachyonFileSystemWrapperTest.java [hunk 12]
```diff
@@ -141,8 +141,8 @@ public void testTachyon() {
 			IOUtils.copy(inStream, writer);
 			String resultString = writer.toString();
 
-			Assert.assertEquals("hello 10\n" +
-					"tachyon 10\n", resultString);
+			Assert.assertEquals("(hello,10)\n" +
```
Comment (StephanEwen) [inline flink-staging/flink-tachyon/src/test/java/org/apache/flink/tachyon/TachyonFileSystemWrapperTest.java]: Followup of adjusted WordCount example

File: flink-staging/flink-tachyon/src/test/java/org/apache/flink/tachyon/HDFSTest.java [hunk 13]
```diff
@@ -110,8 +110,8 @@ public void testHDFS() {
 			IOUtils.copy(inStream, writer);
 			String resultString = writer.toString();
 
-			Assert.assertEquals("hdfs 10\n" +
-					"hello 10\n", resultString);
+			Assert.assertEquals("(hdfs,10)\n" +
```
Comment (StephanEwen) [inline flink-staging/flink-tachyon/src/test/java/org/apache/flink/tachyon/HDFSTest.java]: Followup of adjusted WordCount example

**Comment by StephanEwen** [comment]
I just noticed this test case:
`flink-tests/src/test/scala/org/apache/flink/api/scala/DeltaIterationSanityCheckTest.scala`

The programs in this test are never meant to execute, so we need to swap the `print()` statements with some other lazy output statement.

**Comment by StephanEwen** [comment]
Looks good now. Will merge this.

Thank you for the contribution!

**Comment by mxm** [comment]
Congratulations on your first contribution @nikste!

**Comment by fhueske** [comment]
+1!

**Pull request closed without merging**